src/highdicom/seg/sop.py

```diff
@@ -1932,29 +1905,287 @@ def _omit_empty_frames(
 
         Returns
         -------
-        pixel_array: numpy.ndarray
-            Pixel array with empty frames removed
         plane_positions: List[Optional[highdicom.PlanePositionSequence]]
             Plane positions with entries corresponding to empty frames removed.
         source_image_indices: List[int]
             List giving for each frame in the output pixel array the index of
             the corresponding frame in the original pixel array
+        is_empty: bool
+            Whether the entire image is empty. If so, empty frames should not
+            be omitted.
 
         """
-        non_empty_frames = []
         non_empty_plane_positions = []
 
         # This list tracks which source image each non-empty frame came from
         source_image_indices = []
         for i, (frm, pos) in enumerate(zip(pixel_array, plane_positions)):
-            if frm.sum() > 0:
-                non_empty_frames.append(frm)
+            if np.any(frm):
                 non_empty_plane_positions.append(pos)
                 source_image_indices.append(i)
-        pixel_array = np.stack(non_empty_frames)
-        plane_positions = non_empty_plane_positions
 
-        return (pixel_array, plane_positions, source_image_indices)
+        if len(non_empty_plane_positions) == 0:
+            logger.warning(
+                'Encoding an empty segmentation with "omit_empty_frames" '
+                'set to True. Reverting to encoding all frames since omitting '
+                'all frames is not possible.'
+            )
+            return (plane_positions, list(range(len(plane_positions))), True)
+
+        return (non_empty_plane_positions, source_image_indices, False)
+
+    @staticmethod
+    def _get_segment_array(
+        pixel_array: np.ndarray,
+        segment_number: int,
+        number_of_segments: int,
+        segmentation_type: SegmentationTypeValues,
+        max_fractional_value: int
+    ) -> np.ndarray:
+        """Get segmentation array for a specific segment.
+
+        This is a helper method used during the constructor.
+
+        Parameters
+        ----------
+        pixel_array: numpy.ndarray
+            Full segmentation array containing all segments.
+        segment_number: int
+            The segment of interest.
+        number_of_segments: int
+            Number of segments in the the segmentation.
+        segmentation_type: highdicom.seg.SegmentationTypeValues
+            Desired output segmentation type.
+        max_fractional_value: int
+            Value for scaling FRACTIONAL segmentations.
+
+        """
+        if pixel_array.dtype in (np.float_, np.float32, np.float64):
+            # Based on the previous checks and casting, if we get here the
```

Comment:
This sounds good the checks and casting operations in `_check_and_caste_pixel_array` are a *very* significant part of the time for the entire construction, especially when there is no compression to do. This is because they involve checks on the unique values in the array, overlap etc, on potentially very large array. Consequently we definitely do not want to be doing them once per segment. This does of course introduce some dependencies between this method (which runs in a loop for every segment) and `_check_and_caste_pixel_array` (which runs once before the loop and processes the pixel array). The encapsulation is not perfect. But I think perfect encapsulation is impossible while maintaining reasonable performance so it was not my goal. Grouping similar code into these private methods still significantly improves readability in my opinion. I added this comment here to try and document these dependencies, I also added a further comment in the docstring in c0d9f72 to make this dependency even more explicit